src/tactic/library_search.lean

```diff
@@ -76,29 +76,75 @@ meta def library_defs (hs : name) : tactic (list decl_data) :=
 do env ← get_env,
    return $ env.decl_filter_map (process_declaration hs)
 
-meta def apply_and_solve (discharger : tactic unit) (e : expr) :=
+/-- Fail if the target contains a metavariable. -/
+meta def no_mvars_in_target : tactic unit :=
+expr.has_meta_var <$> target >>= guardb ∘ bnot
+
+/-
+Fail unless the goal is "safe", i.e. propositional, and does not include any metavariables.
+If a goal is "safe", discharging it does not affect the provability of any other goals.
+-/
+meta def safe : tactic unit :=
```

Comment:
This and `no_mvars_in_target` can probably move to `tactic.core`, although `safe` should be renamed.

Reply:
`tactic.core` already has the much more ambitious `terminal_goal`, which tries to establish that no other goals can depend on the value of the given goal (propositional and subsingleton goals are automatically terminal, and after that we search the types of other metavariables in the result for the current metavariable). I could just use that.

Reply:
Done.